src/net.cpp

```diff
@@ -408,9 +409,32 @@ CNode* CConnman::FindNode(const CService& addr)
     return nullptr;
 }
 
-bool CConnman::AlreadyConnectedToAddress(const CAddress& addr)
+bool CConnman::AlreadyConnectedToAddress(const CAddress& addr, ConnectionType conn_type, bool log)
 {
-    return FindNode(static_cast<CNetAddr>(addr)) || FindNode(addr.ToStringAddrPort());
+    if (Params().IsTestChain()) return false; // allow bilateral connections on test chains
+    const CService service{MaybeFlipIPv6toCJDNS(addr)};
+    const bool inbound{conn_type == ConnectionType::INBOUND};
+    CNode* pnode{nullptr};
+    if (inbound) {
+        // For accepting an inbound connection, check all existing connections by addr only.
+        pnode = FindNode(static_cast<CNetAddr>(service));
+    } else {
+        // For making an outbound connection, check the existing outbound connections
+        // by addr and port, and if none, then inbound connections by addr only.
+        pnode = FindNode(service.ToStringAddrPort(), /*inbound=*/false);
```

Comment:
Can use `CConnman::FindNode(const CService& addr)` here?
```suggestion
        pnode = FindNode(service, /*inbound=*/false);
```

but then we will not be comparing against `CNode::m_addr_name`. I wonder if we should compare against both `CNode::addr` and `CNode::m_addr_name`?

Reply:
Looks good. Can `CNode::addr` ever be invalid?